controller/volume_controller.go

```diff
@@ -1921,11 +1923,13 @@ func (c *VolumeController) reconcileVolumeSize(v *longhorn.Volume, e *longhorn.E
 }
 
 func (c *VolumeController) canInstanceManagerLaunchReplica(r *longhorn.Replica) (bool, error) {
+	log := c.logger.WithField("replica", r.Name)
 	nodeDown, err := c.ds.IsNodeDownOrDeleted(r.Spec.NodeID)
 	if err != nil {
 		return false, errors.Wrapf(err, "fail to check IsNodeDownOrDeleted %v", r.Spec.NodeID)
 	}
 	if nodeDown {
+		log.Warnf("Cannot launch replica due to node %v is NotReady or Deleted", r.Spec.NodeID)
```

Comment:
This log might be misleading because it's not an action but rather a query. Typically, for query functions/methods (such as is/get/check/...), I would suggest avoiding logging inside them except using lower log levels (ex: debug, trace..). Instead, it's better to utilize returned variables including error, allowing the caller to determine the reason. This approach could also prevent unexpected log flooding in case it will be used frequently.

Since the only caller already dumps the warning log as below, do you still this we need this log here in this method?

[/controller/volume_controller.go#L1691-L1725](https://github.com/longhorn/longhorn-manager/blob/7496347232f4f98ad51a1c5b35c2599db896499a/controller/volume_controller.go#L1691-L1725)
```go
		canIMLaunchReplica, err := c.canInstanceManagerLaunchReplica(r)
		if err != nil {
			return err
		}
		if canIMLaunchReplica {
			if r.Spec.FailedAt == "" && r.Spec.Image == v.Status.CurrentImage {
				if r.Status.CurrentState == longhorn.InstanceStateStopped {
					r.Spec.DesireState = longhorn.InstanceStateRunning
				}
			}
		} else {
			// wait for IM is starting when volume is upgrading
			if c.isVolumeUpgrading(v) {
				continue
			}

			// If the engine isn't attached or the node goes down, the replica can be marked as failed.
			// In the attached mode, we can determine whether the replica can fail by relying on the data plane's connectivity status.
			nodeDeleted, err := c.ds.IsNodeDeleted(r.Spec.NodeID)
			if err != nil {
				return err
			}

			if v.Status.State != longhorn.VolumeStateAttached || nodeDeleted {
				msg := fmt.Sprintf("Replica %v is marked as failed since the volume %v is not attached because the instance manager is unable to launch the replica", r.Name, v.Name)
				if nodeDeleted {
					msg = fmt.Sprintf("Replica %v is marked as failed since the node %v is deleted.", r.Name, r.Spec.NodeID)
				}
				log.WithField("replica", r.Name).Warn(msg)
				if r.Spec.FailedAt == "" {
					r.Spec.FailedAt = c.nowHandler()
				}
				r.Spec.DesireState = longhorn.InstanceStateStopped
			}
		}
```

Reply:
Hi @innobead, 

I checked again with the whole logic, and the above error message is only for the node deleted.
Could we use the `c.ds.IsNodeDownOrDeleted(r.Spec.NodeID)` instead `c.ds.IsNodeDeleted(r.Spec.NodeID)` and make minor logging improvements here?